src/databricks/labs/ucx/hive_metastore/federation.py

```diff
@@ -0,0 +1,112 @@
+import collections
+import logging
+
+from databricks.sdk import WorkspaceClient
+from databricks.sdk.errors import AlreadyExists, NotFound, BadRequest
+from databricks.sdk.service.catalog import (
+    ConnectionType,
+    ConnectionInfo,
+    SecurableType,
+    Privilege,
+    PermissionsChange,
+)
+
+from databricks.labs.blueprint.installation import Installation
+from databricks.labs.ucx.account.workspaces import WorkspaceInfo
+from databricks.labs.ucx.config import WorkspaceConfig
+from databricks.labs.ucx.hive_metastore import ExternalLocations
+
+
+logger = logging.getLogger(__name__)
+
+class HiveMetastoreFederationEnabler:
+    def __init__(self, installation: Installation):
+        self._installation = installation
+
+    def enable(self):
+        config = self._installation.load(WorkspaceConfig)
+        config.enable_hms_federation = True
+        self._installation.save(config)
+
+class HiveMetastoreFederation:
+    def __init__(
+        self,
+        workspace_client: WorkspaceClient,
+        external_locations: ExternalLocations,
+        workspace_info: WorkspaceInfo,
+    ):
+        self._workspace_client = workspace_client
+        self._external_locations = external_locations
+        self._workspace_info = workspace_info
+
+    def run(self):
+        name = self._workspace_info.current()
+        connection_info = self._get_or_create_connection(name)
+        try:
+            self._workspace_client.catalogs.create(
+                name=connection_info.name,
+                connection_name=connection_info.name,
+                options={"authorized_paths": self._get_authorized_paths()},
+            )
+        except BadRequest as err:
+            if err.error_code == 'CATALOG_ALREADY_EXISTS':
+                logger.info(f'Catalog {connection_info.name} already exists')
+                for catalog_info in self._workspace_client.catalogs.list():
+                    if catalog_info.name == connection_info.name:
+                        return catalog_info
+            raise err
+
+    def _get_or_create_connection(self, name: str) -> ConnectionInfo:
+        try:
+            return self._workspace_client.connections.create(
+                name=name,
+                connection_type=ConnectionType.HIVE_METASTORE,  # needs SDK change
+                options={"builtin": "true"},
+            )
+        except AlreadyExists:
+            for connection in self._workspace_client.connections.list():
+                if connection.name == name:
+                    return connection
+        raise NotFound(f'Connection {name} not found')
+
+    def _get_authorized_paths(self) -> str:
```

Comment:
Plural "paths" usually means a collection type

```suggestion
    def _get_authorized_paths(self) -> list[str]:
```

Reply:
Note that this method does more than just getting the authorized paths, it also sets the missing permissions if needed, update the name accordingly or separate out the two functionalities:
1. Getting paths
2. Setting missing permissions